addons/web/static/src/views/fields/dynamicplaceholder_hook.js

```diff
@@ -2,60 +2,127 @@
+            element.value = newValue;
+
+            // -1 to take the removal of the trigger key char into account
+            rangeIndex += dynamicPlaceholder.length - 1;
+            element.setSelectionRange(rangeIndex, rangeIndex);
+            element.removeAttribute("data-oe-dynamic-placeholder-range-index");
         }
     };
-}
+    const onDynamicPlaceholderClose = function () {
+        elementRef?.el.focus();
+    };
+
+    /**
+     * Open a Model Field Selector which can select fields to create a dynamic
+     * placeholder string in the Input with or without a default text value.
+     *
+     * @public
+     * @param {Object} options
+     * @param {function} options.validateCallback
+     * @param {function} options.closeCallback
+     * @param {function} [options.positionCallback]
+     */
+    async function open(options) {
+        if (!baseModel) {
+            return notification.add(
+                ownerField.env._t(
+                    "You need to select a baseModel before opening the dynamic placeholder selector."
+                ),
+                { type: "danger" }
+            );
+        }
 
+        dynamicPlaceholderChain = await modelField.loadChain(baseModel, "");
+        popover.add(
+            elementRef?.el,
+            ModelFieldSelectorPopover,
+            {
+                chain: dynamicPlaceholderChain,
+                update: update,
+                validate: options.validateCallback,
+                showSearchInput: true,
+                isDebugMode: true,
+                needDefaultValue: true,
+                loadChain: modelField.loadChain,
+                filter: (model) => !["one2many", "boolean", "many2many"].includes(model.type),
+            },
+            {
+                closeOnClickAway: true,
+                onClose: options.closeCallback,
+                onPositioned: options.positionCallback,
+            }
+        );
+    }
+
+    async function onKeydownListener(ev) {
+        const element = elementRef?.el;
+        if (ev.target === element && ev.key === TRIGGER_KEY) {
+            const currentRangeIndex = element.selectionStart;
+            // +1 to take the trigger key char into account
+            element.setAttribute("data-oe-dynamic-placeholder-range-index", currentRangeIndex + 1);
+            await open({
+                validateCallback: onDynamicPlaceholderValidate,
+                closeCallback: onDynamicPlaceholderClose,
+            });
+        }
+    }
+    function addTriggerKeyListener() {
+        document.addEventListener("keydown", onKeydownListener);
+    }
+    function removeTriggerKeyListener() {
+        document.removeEventListener("keydown", onKeydownListener);
+    }
+    function refreshBaseModel() {
+        const recordData = ownerField.props.record.data;
+        baseModel = recordData?.mailing_model_real || recordData.model;
+    }
 
+    return {
+        TRIGGER_KEY: TRIGGER_KEY,
```

Comment:
We might be able to make further simplification by removing this 4 lines and moving the use of `onMounted`, `useEffect` and `onWillUnmount` inside this hook.

```js
function useDynamicPlaceholder(elementRef) {
  // this replaces the use of `onMounted` and `onWillUnmount` to set/remove the listener.
  useExternalListener(document, 'keydown', onKeydownListener);

  // This will properly call refreshBaseModel during mounted as well.
  useEffect(refreshBaseModel);

  ...
}
```

Doing this, you can remove the use of `onMounted`, `onWillUnmount` and `useEffect` in the dependent components because they are handled in the `useDynamicPlaceholder` hook.

I'm also asking to delete the TRIGGER_KEY export because it doesn't seem to be used outside. But if I'm mistaken, keep it exported.

Reply:
Oh, you can keep exposing `refreshBaseModel` because it's used in html_field.

Reply:
> We might be able to make further simplification by removing this 4 lines and moving the use of onMounted, useEffect and onWillUnmount inside this hook.

> Doing this, you can remove the use of onMounted, onWillUnmount and useEffect in the dependent components because they are handled in the useDynamicPlaceholder hook.

The reason the  `onMounted`, `onWillUnmount` and `useEffect` are not in the hook, is because their are not used/ diferent in the HTML_Field case.

> I'm also asking to delete the TRIGGER_KEY export because it doesn't seem to be used outside. But if I'm mistaken, keep it exported.

:+1:

Reply:
> The reason the `onMounted`, `onWillUnmount` and `useEffect` are not in the hook, is because their are not used/ diferent in the HTML_Field case.

I see. Maybe we can use the trick in the place where it's applicable. e.g. in char_field, we can instead do something like this:

```js
class CharField extends Component {
  setup() {
    const dynamicPlaceholder = useDynamicPlaceholder(this.input);
    useExternalListener(document, 'keydown', dynamicPlaceholder.onKeydownListener);
    useEffect(dynamicPlaceholder.refreshBaseModel);
    ...
  }
}
```

Reply:
This is already the case right ?
Except it's conditionally bound to `this.props.dynamicPlaceholder`.

Reply:
changes done ( see https://github.com/odoo-dev/odoo/pull/1815 )